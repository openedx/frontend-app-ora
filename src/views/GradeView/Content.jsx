import React from 'react';
import { v4 as uuid } from 'uuid';

import { useIntl } from '@edx/frontend-platform/i18n';

import {
  usePrompts,
  useResponseData,
  useEffectiveGradeStep,
} from 'hooks/app';
import { stepNames } from 'constants';
import apiMessages from 'data/services/lms/messages';

import FileUpload from 'components/FileUpload';
import Prompt from 'components/Prompt';
import TextResponse from 'components/TextResponse';
import messages from './messages';

const Content = () => {
  const prompts = usePrompts();
  const response = useResponseData();
  const effectiveGradeStep = useEffectiveGradeStep();
  const { formatMessage } = useIntl();
  const stepLabel = formatMessage(apiMessages[effectiveGradeStep]);
  return (
    <div>
      <strong>{formatMessage(messages.aboutYourGrade)}</strong>
      <p>
        {formatMessage(messages.finalGradeInfo, { step: stepLabel })}
        {effectiveGradeStep === stepNames.peer && (
          <>
            <br />
            {formatMessage(messages.peerAsFinalGradeInfo)}
          </>
        )}
      </p>
      <div>
        {
          prompts.map((prompt, index) => (
            <div key={uuid()}>
              <Prompt prompt={prompt} defaultOpen={false} />
              {response.textResponses[index] && <TextResponse response={response.textResponses[index]} />}
            </div>
          ))
        }
        <FileUpload isReadOnly uploadedFiles={response.uploadedFiles} defaultCollapsePreview />
      </div>
    </div>
  );
};

Content.defaultProps = {};
Content.propTypes = {};

export default Content;
