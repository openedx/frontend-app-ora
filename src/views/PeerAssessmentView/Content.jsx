import React from 'react';

import {
  usePrompts,
  useResponseData,
} from 'data/services/lms/hooks/selectors';

import Prompt from 'components/Prompt';
import TextResponse from 'components/TextResponse';
import FileUpload from 'components/FileUpload';

const AssessmentContent = () => {
  const prompts = usePrompts();
  const response = useResponseData();
  return (
    <div>
      {React.Children.toArray(
        prompts.map((prompt, index) => (
          <div>
            <Prompt prompt={prompt} />
            <TextResponse response={response.textResponses[index]} />
          </div>
        )),
      )}
      <FileUpload isReadOnly uploadedFiles={response.uploadedFiles} />
    </div>
  );
};

export default AssessmentContent;
