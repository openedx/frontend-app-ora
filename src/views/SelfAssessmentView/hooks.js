import { useORAConfigData, usePageData } from 'data/services/lms/hooks/selectors';

const useAssessmentViewHooks = () => {
  const pageData = usePageData();
  const oraConfigData = useORAConfigData();

  return {
    submission: pageData?.submission,
    oraConfigData,
  };
};

export default useAssessmentViewHooks;
