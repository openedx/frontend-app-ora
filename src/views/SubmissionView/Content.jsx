import React from 'react';
import PropTypes from 'prop-types';

import { Icon } from '@edx/paragon';
import { CheckCircle } from '@edx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import {
  usePrompts,
  useSubmissionConfig,
  useStepState,
} from 'data/services/lms/hooks/selectors';
import { stepNames, stepStates } from 'data/services/lms/constants';

import FileUpload from 'components/FileUpload';
import Instructions from 'components/Instructions';
import Prompt from 'components/Prompt';
import TextResponse from 'components/TextResponse';
import TextResponseEditor from 'components/TextResponseEditor';
import StatusAlert from 'components/StatusAlert';

import messages from './messages';

const SubmissionContent = ({
  textResponses,
  uploadedFiles,
}) => {
  const submissionConfig = useSubmissionConfig();

  const stepState = useStepState({ step: stepNames.submission });
  const isReadOnly = stepState === stepStates.completed;
  const prompts = usePrompts();
  const { formatMessage } = useIntl();
  const createTextResponse = (index) => (isReadOnly
    ? <TextResponse response={textResponses.value[index]} />
    : (
      <TextResponseEditor
        submissionConfig={submissionConfig}
        value={textResponses.value[index]}
        onChange={textResponses.onChange(index)}
      />
    ));

  return (
    <div>
      <div className="d-flex justify-content-between">
        <h2 className="mb-4">{formatMessage(messages.yourResponse)}</h2>
        {(!isReadOnly && textResponses.draftSaved) && (
          <p className="d-flex text-primary">
            <Icon src={CheckCircle} />
            {formatMessage(messages.draftSaved)}
          </p>
        )}
      </div>

      <StatusAlert />
      <Instructions />

      {prompts.map((prompt, index) => (
        // eslint-disable-next-line react/no-array-index-key
        <div key={index}>
          <Prompt prompt={prompt} />
          {createTextResponse(index)}
        </div>
      ))}

      <FileUpload
        onDeletedFile={uploadedFiles.onDeletedFile}
        onFileUploaded={uploadedFiles.onFileUploaded}
        uploadedFiles={uploadedFiles.value}
        isReadOnly={isReadOnly}
      />
    </div>
  );
};

SubmissionContent.propTypes = {
  textResponses: PropTypes.shape({
    value: PropTypes.arrayOf(PropTypes.string).isRequired,
    onChange: PropTypes.func.isRequired,
    draftSaved: PropTypes.bool.isRequired,
  }).isRequired,
  uploadedFiles: PropTypes.shape({
    value: PropTypes.arrayOf(PropTypes.shape({
      fileDescription: PropTypes.string,
      fileName: PropTypes.string,
      fileSize: PropTypes.number,
    })),
    onDeletedFile: PropTypes.func.isRequired,
    onFileUploaded: PropTypes.func.isRequired,
  }).isRequired,
};

export default SubmissionContent;
