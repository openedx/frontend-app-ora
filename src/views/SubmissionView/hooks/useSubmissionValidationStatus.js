import { useCallback } from 'react';
import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';

import {
  useSubmissionConfig,
} from 'hooks/app';

export const stateKeys = StrictDict({
  promptStatuses: 'promptStatuses',
  submissionTriggered: 'submissionTriggered',
  fileUploadIsRequired: 'fileUploadIsRequired',
});

const useSubmissionStatuses = (textResponses, uploadedFiles) => {
  const submissionConfig = useSubmissionConfig();

  // holds the status of each text field
  const [promptStatuses, setPromptStatuses] = useKeyedState(
    stateKeys.promptStatuses,
    textResponses.reduce((obj, val, i) => ({ ...obj, [i]: true }), {}),
  );
  const [submissionTriggered, setSubmissionTriggered] = useKeyedState(stateKeys.submissionTriggered, false);
  const [fileUploadIsRequired, setFileUploadIsRequired] = useKeyedState(stateKeys.fileUploadIsRequired, false);

  // this function is to be called when response submit button is clicked
  const validateBeforeConfirmation = useCallback(() => {
    setSubmissionTriggered(true);
    const retrievePromptsSubmissionStatus = textResponses.reduce((obj, val, i) => ({ ...obj, [i]: val }), {});
    let containsInvalidPrompts = false;
    if (submissionConfig.textResponseConfig.required) {
      for (let i = 0; i < textResponses.length; i++) {
        retrievePromptsSubmissionStatus[i] = !!textResponses[i];
      }
      containsInvalidPrompts = Object.values(retrievePromptsSubmissionStatus).some((val) => !val);
    }
    const calcFileUploadIsRequired = submissionConfig.fileResponseConfig.required && !uploadedFiles.length;
    setFileUploadIsRequired(calcFileUploadIsRequired);
    if (containsInvalidPrompts || (submissionConfig.fileResponseConfig.required && !uploadedFiles.length)) {
      setPromptStatuses(retrievePromptsSubmissionStatus);
      return false;
    }

    return true;
  }, [
    textResponses,
    uploadedFiles,
    submissionConfig,
    setPromptStatuses,
    setSubmissionTriggered,
    setFileUploadIsRequired,
  ]);

  return {
    validateBeforeConfirmation,
    setSubmissionTriggered,
    submissionTriggered,
    setPromptStatuses,
    promptStatuses,
    fileUploadIsRequired,
  };
};

export default useSubmissionStatuses;
