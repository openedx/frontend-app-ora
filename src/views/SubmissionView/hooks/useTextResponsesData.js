import { useCallback, useEffect } from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';
import {
  useFinishLater,
  useHasSubmitted,
  useSaveDraftResponse,
  useTextResponses,
} from 'hooks/app';
import { MutationStatus } from 'constants/index';

export const stateKeys = StrictDict({
  textResponses: 'textResponses',
  isDirty: 'isDirty',
  hasSaved: 'hasSaved',
  lastChanged: 'lastChanged',
});

const useTextResponsesData = ({ setHasSavedDraft }) => {
  const textResponses = useTextResponses();
  const hasSubmitted = useHasSubmitted();
  const [hasSaved, setHasSaved] = useKeyedState(stateKeys.hasSaved, false);
  const [lastChanged, setLastChanged] = useKeyedState(stateKeys.lastChanged, null);
  const [isDirty, setIsDirty] = useKeyedState(stateKeys.isDirty, false);
  const [value, setValue] = useKeyedState(stateKeys.textResponses, textResponses);

  const saveResponseMutation = useSaveDraftResponse();
  const finishLaterMutation = useFinishLater();

  const saveResponse = useCallback(() => {
    setIsDirty(false);
    return saveResponseMutation.mutateAsync({ textResponses: value }).then(() => {
      setHasSavedDraft(true);
    });
  }, [
    saveResponseMutation,
    value,
    setIsDirty,
    setHasSavedDraft,
  ]);

  const finishLater = useCallback(() => {
    setIsDirty(false);
    return finishLaterMutation.mutateAsync({ textResponses: value });
  }, [setIsDirty, finishLaterMutation, value]);

  const onChange = useCallback((index) => (textResponse) => {
    setValue(oldResponses => {
      const out = [...oldResponses];
      out[index] = textResponse;
      return out;
    });
    setIsDirty(true);
    setLastChanged(Date.now());
  }, [
    setValue,
    setIsDirty,
    setLastChanged,
  ]);

  useEffect(() => {
    const interval = setInterval(() => {
      if (!lastChanged) {
        return;
      }
      const timeSinceChange = Date.now() - lastChanged;
      if (isDirty && timeSinceChange > 2000) {
        saveResponse();
        if (!hasSaved) {
          setHasSaved(true);
        }
      }
    }, 2000);
    return () => clearInterval(interval);
  }, [
    isDirty,
    hasSubmitted,
    hasSaved,
    setHasSaved,
    lastChanged,
    saveResponse,
  ]);

  return {
    textResponses: value,
    onUpdateTextResponse: onChange,
    isDraftSaved: saveResponseMutation.status === MutationStatus.success && !isDirty,
    saveResponse,
    saveResponseStatus: saveResponseMutation.status,
    finishLater,
    finishLaterStatus: finishLaterMutation.status,
  };
};

export default useTextResponsesData;
