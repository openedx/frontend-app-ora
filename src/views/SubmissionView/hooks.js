import { useEffect, useReducer } from 'react';

import { useORAConfigData, usePageData } from 'data/services/lms/hooks/selectors';

import { submitResponse, saveResponse, uploadFiles } from 'data/services/lms/hooks/actions';
import { MutationStatus } from 'data/services/lms/constants';

const useSubmissionViewHooks = () => {
  const submitResponseMutation = submitResponse();
  const saveResponseMutation = saveResponse();
  const uploadFilesMutation = uploadFiles();
  const pageData = usePageData();
  const oraConfigData = useORAConfigData();

  const [submission, dispatchSubmission] = useReducer(
    (state, payload) => ({ ...state, isDirty: true, ...payload }),
    { ...pageData?.submission, isDirty: false },
  );

  useEffect(() => {
    // a workaround to update the submission state when the pageData changes
    if (pageData?.submission) {
      dispatchSubmission({ ...pageData.submission, isDirty: false });
    }
  }, [pageData?.submission]);

  const onTextResponseChange = (index) => (textResponse) => {
    dispatchSubmission({
      response: {
        ...submission.response,
        textResponses: [
          ...submission.response.textResponses.slice(0, index),
          textResponse,
          ...submission.response.textResponses.slice(index + 1),
        ],
      },
    });
  };

  const onFileUploaded = (args) => {
    const fileUploads = uploadFilesMutation.mutate(args);
    dispatchSubmission({ response: { ...submission.response, fileUploads } });
  };

  const submitResponseHandler = () => {
    dispatchSubmission({ isDirty: false });
    submitResponseMutation.mutate(submission);
  };

  const saveResponseHandler = () => {
    dispatchSubmission({ isDirty: false });
    saveResponseMutation.mutate(submission);
  };

  return {
    submitResponseHandler,
    submitResponseStatus: submitResponseMutation.status,
    saveResponseHandler,
    saveResponseStatus: saveResponseMutation.status,
    draftSaved: saveResponseMutation.status === MutationStatus.success && !submission.isDirty,
    pageData,
    oraConfigData,
    submission,
    dispatchSubmission,
    onTextResponseChange,
    onFileUploaded,
  };
};

export default useSubmissionViewHooks;
