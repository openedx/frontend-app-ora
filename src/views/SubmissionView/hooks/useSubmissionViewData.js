import React from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';

import {
  useGlobalState,
  useRubricConfig,
  useSubmitResponse,
  useSetHasSubmitted,
  useHasSubmitted,
  useRefreshPageData,
  useSetResponse,
  useResponse,
} from 'hooks/app';
import {
  useRefreshUpstream,
} from 'hooks/modal';
import { stepStates, stepNames } from 'constants/index';

import useTextResponsesData from './useTextResponsesData';
import useUploadedFilesData from './useUploadedFilesData';

const stateKeys = StrictDict({
  hasSavedDraft: 'hasSavedDraft',
});

const useSubmissionViewData = () => {
  const [hasSavedDraft, setHasSavedDraft] = useKeyedState(stateKeys.hasSavedDraft, false);
  const hasSubmitted = useHasSubmitted();
  const setHasSubmitted = useSetHasSubmitted();
  const submitResponseMutation = useSubmitResponse();
  const rubricConfig = useRubricConfig();
  const globalState = useGlobalState({ step: stepNames.submission });
  const refreshPageData = useRefreshPageData();
  const setResponse = useSetResponse();
  const response = useResponse();
  const refreshUpstream = useRefreshUpstream();
  const stepState = hasSubmitted ? stepStates.submitted : globalState.stepState;

  const {
    textResponses,
    onUpdateTextResponse,
    saveResponse,
    saveResponseStatus,
    finishLater,
    finishLaterStatus,
  } = useTextResponsesData({ setHasSavedDraft });

  const {
    uploadedFiles,
    onFileUploaded,
    onDeletedFile,
  } = useUploadedFilesData();

  const submitResponseHandler = React.useCallback(() => {
    submitResponseMutation.mutateAsync({
      textResponses,
      uploadedFiles,
    }).then(() => {
      setResponse({ textResponses, uploadedFiles });
      setHasSubmitted(true);
      refreshPageData();
      refreshUpstream();
    });
  }, [ // eslint-disable-line react-hooks/exhaustive-deps
    setHasSubmitted,
    submitResponseMutation,
    textResponses,
    uploadedFiles,
  ]);

  return {
    actionOptions: {
      finishLater,
      finishLaterStatus,
      saveResponse,
      saveResponseStatus,
      submit: submitResponseHandler,
      submitStatus: submitResponseMutation.status,
      hasSubmitted,
    },
    response: hasSubmitted
      ? response
      : { textResponses, uploadedFiles },
    hasSubmitted,
    onUpdateTextResponse,
    isDraftSaved: hasSavedDraft,
    onDeletedFile,
    onFileUploaded,
    showRubric: rubricConfig.showDuringResponse,
    isReadOnly: stepState === stepStates.done || hasSubmitted,
  };
};

export default useSubmissionViewData;
