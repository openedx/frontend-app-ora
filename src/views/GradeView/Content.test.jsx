import { shallow } from '@edx/react-unit-test-utils';

import { usePrompts, useResponseData, useEffectiveGradeStep } from 'hooks/app';
import { stepNames } from 'constants/index';

import Content from './Content';

jest.mock('hooks/app', () => ({
  usePrompts: jest.fn(),
  useResponseData: jest.fn(),
  useEffectiveGradeStep: jest.fn(),
}));
jest.mock('components/FileUpload', () => 'FileUpload');
jest.mock('components/Prompt', () => 'Prompt');
jest.mock('components/TextResponse', () => 'TextResponse');

describe('<Content />', () => {
  it('render without prompt and effectiveGradeStep is not peer', () => {
    usePrompts.mockReturnValue([]);
    useResponseData.mockReturnValue({ textResponses: [], uploadedFiles: [] });
    useEffectiveGradeStep.mockReturnValue(stepNames.self);
    const wrapper = shallow(<Content />);
    expect(wrapper.snapshot).toMatchSnapshot();
  });

  it('render with prompt, textResponse and effectiveGradeStep is peer', () => {
    usePrompts.mockReturnValue(['prompt1', 'prompt2']);
    useResponseData.mockReturnValue({
      textResponses: ['text response 1', 'text response 2'],
      uploadedFiles: ['upload'],
    });
    useEffectiveGradeStep.mockReturnValue(stepNames.peer);
    const wrapper = shallow(<Content />);
    expect(wrapper.snapshot).toMatchSnapshot();
  });
});
