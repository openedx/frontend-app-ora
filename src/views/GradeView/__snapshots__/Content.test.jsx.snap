// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Content /> render with prompt, textResponse and effectiveGradeStep is peer 1`] = `
<div>
  <strong>
    About your grade: 
  </strong>
  <p>
    Your grade is based on your Peer assessment score for this problem. Other assessments don't count towards your final score.
    <Fragment>
      <br />
      Only the required number of peer grades will counted against your final grade. The others are shown, but are not included in your grade calculation
    </Fragment>
  </p>
  <div>
    <div
      key="some_uuid"
    >
      <Prompt
        onToggle={
          Object {
            "useCallback": Object {
              "cb": [Function],
              "prereqs": Array [
                false,
              ],
            },
          }
        }
        open={false}
        prompt="prompt1"
      />
      <h3
        className="pt-2"
      >
        Your response
      </h3>
      <TextResponse
        response="text response 1"
      />
    </div>
    <div
      key="some_uuid"
    >
      <Prompt
        onToggle={
          Object {
            "useCallback": Object {
              "cb": [Function],
              "prereqs": Array [
                false,
              ],
            },
          }
        }
        open={false}
        prompt="prompt2"
      />
      <h3
        className="pt-2"
      >
        Your response
      </h3>
      <TextResponse
        response="text response 2"
      />
    </div>
    <FileUpload
      defaultCollapsePreview={true}
      hideHeader={true}
      isReadOnly={true}
      uploadedFiles={
        Array [
          "upload",
        ]
      }
    />
  </div>
</div>
`;

exports[`<Content /> render without prompt and effectiveGradeStep is not peer 1`] = `
<div>
  <strong>
    About your grade: 
  </strong>
  <p>
    Your grade is based on your Self assessment score for this problem. Other assessments don't count towards your final score.
  </p>
  <div>
    <FileUpload
      defaultCollapsePreview={true}
      hideHeader={true}
      isReadOnly={true}
      uploadedFiles={Array []}
    />
  </div>
</div>
`;
